File: subprojects/docs/src/snippets/userguide/plugins/readOnlyManagedProperty/java/buildSrc/src/main/java/Download.java
import org.gradle.api.DefaultTask;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.TaskAction;

import java.net.URI;

public abstract class Download extends DefaultTask {
    // Use an abstract getter method
    @Input
    abstract Property<URI> getUri();

    @TaskAction
    void run() {
        // Use the `uri` property
        System.out.println("Downloading " + getUri().get());
    }
}

File: subprojects/docs/src/snippets/userguide/plugins/readOnlyManagedProperty/kotlin/buildSrc/src/main/kotlin/Download.kt
import org.gradle.api.DefaultTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.TaskAction
import org.gradle.api.provider.Property
import java.net.URI

abstract class Download : DefaultTask() {
    // Use an abstract val
    @get:Input
    abstract val uri: Property<URI>

    @TaskAction
    fun run() {
        // Use the `uri` property
        println("Downloading ${uri.get()}")
    }
}
